### DISCUSSION TITLE
Will/Should Firebase Plugins be implemented natively in Dart?

### DISCUSSION THREAD

**Discussion by Jonas-Sander** [firebase/flutterfire#3359] [General] [answered]
The current plugins are all calling native implementations via [platform channels](https://flutter.dev/docs/development/platform-integration/platform-channels).

I would love to have a discussion on:
- Why are the current plugins all implemented via platform channels?
- Why should or shouldn't these Plugins be implemented natively via Dart in the future?

This discussion was already opened once here as a ticket but there was no answer from the FlutterFire team:   
https://github.com/FirebaseExtended/flutterfire/issues/2191

Efforts made by the community for native implementations:
- https://github.com/cachapa/firedart
- https://github.com/fluttercommunity/firebase_dart_sdk

**Comment by Jonas-Sander** [comment]
One big plus for me as a user of these plugins would be to have easy integration tests with services like Firestore.

Currently the only way (that I know of) to run tests using Firestore is via Flutter driver tests.  
It is not possible (at least the last time I tried) to call e.g. Firestore using `package:test` or `flutter:test` as the Plugin doesn't have an e.g. Android/iOS implementation to call via the platform channels.   
(I'm are ignoring `package:e2e` here, which still needs a device).  

Additionally, while I already have seen improvements with recents efforts of the team (thanks guys ❤️) there may still be a benefit to having a Plugin completely written in Dart in regards to build time.

**Comment by Jonas-Sander** [comment]
> Why are the current plugins all implemented via platform channels?

I'd guess this was probably just the quickest way to get started with the least amount of work - still may be interesting to hear if native Dart was considered in the past.

**Comment by slightfoot** [comment]
Bringing in Raz creator of the firebase_dart_sdk above.

/cc @long1eu

**Comment by Salakar** [comment, marked as answer]
I would love this and I have thought about this a lot (like a lot, a lot) so here's my 2 cents on this (but its **not an 'official opinion'** or anything like that);

While it certainly would be possible have some of the plugins in Dart there's quite a few that just won't be possible to do well (or at all) without using underlying native APIs, some examples;

 - Messaging (requires FCM on Android & APNs on iOS/macOS)
 - Dynamic Links (requires native deep linking functionality)
 - Crashlytics (would work for Dart errors, but what about native errors crashes)
 - AdMob (not fully looked into this one, but iOS has Ads personalisation consent SDKs that need interfacing to, Android has similar such as EU consent).

Then you have ones like Firestore that have local first and offline querying capability built into the current SDKs (which is a huge amount of work to implement and to get right, I think) and that each of the current native Firestore SDKs are optimised well for each of their platforms. It could be done though, but it's an impractical amount of work I think.

---

_Option B)_ You could say 'ok how about _some_ of the plugins built in Dart', which leads to problems like synchronising state between 2 SDKs (Dart and non Dart ones), e.g. imagine Auth is all Dart, but Firestore isn't, to get Firestore to work correctly on native with Authentication you'd have still use the native Auth SDK and somehow sync Authentication between the pure Dart auth SDK and the native Firebase Auth SDK so that the Firestore Native SDK still functions with auth.

**Reply by Salakar** [reply]
I would like to see a Dart Firebase Admin SDK though 🤓

**Reply by Jonas-Sander** [reply]
Thanks for the detailed answer. Makes total sense :)

When I personally think about it Auth and Firestore would be the 2 plugins where I would see the most benefit- although you already showed the problems/trade-offs.

> I would like to see a Dart Firebase Admin SDK though 🤓

Yesssss 😍

I think the problem there was (at least some time ago) that some of the Admin-APIs used by the official SDKs were/are not officially released. Not sure though.   
Personally I'm using Dart also for the backend via Cloud Run and the [googleapis](https://pub.dev/packages/googleapis) package is... lets just say not that great in some cases.

**Reply by Jonas-Sander** [reply]
> I would love this and I have thought about this a lot (like a lot, a lot) so here's my 2 cents on this (but its not an 'official opinion' or anything like that);

Would still accept this as the answer for now - if this is alright with you?

**Reply by Salakar** [reply]
Sure

**Reply by orestesgaolin** [reply]
> which is a huge amount of work to implement and to get right, I think

Firestore and Auth have already been ported to Dart - take a look here https://github.com/fluttercommunity/firebase_dart_sdk/tree/develop/cloud_firestore/cloud_firestore_dart cc'ing @long1eu :)

**Reply by Salakar** [reply]
Aye I saw, this is what I meant in general about offline/local support: https://github.com/fluttercommunity/firebase_dart_sdk/tree/develop/cloud_firestore/cloud_firestore_vm/lib/src/firebase/firestore/local

Thats no small amount of code 😅

**Reply by Ehesp** [reply]
Also to add, the Dart implementation would have to be absolutely battle tested and fully managed in case of critical issues / security etc. As a community project this is a massive amount of work.

**Reply by Ehesp** [reply]
@orestesgaolin just found myself discussing this point again elsewhere. I think another point to take into consideration is the upkeep of such a library. The https://github.com/fluttercommunity/firebase_dart_sdk repository now hasn't had an update in 8 months - such a library needs constant attention and upkeep to be viable.

**Reply by orestesgaolin** [reply]
Agree, and I think it's impossible either without huge community support or company standing behind that like Invertase with this very project.

**Comment by Salakar** [comment]
Would like to add as well that we are making all plugins in FlutterFire federated as part of the roadmap. So the `platform_interface` packages can be reused on any efforts relating to this in future if need be.

**Reply by orestesgaolin** [reply]
I wonder if it's possible to follow the `platform_interface` api but still write the code as a package, not the plugin. So for instance the alternative Dart implementation would be written as Dart package, but at the same time would be used in place of e.g. iOS Firestore plugin. Is such thing even possible?

**Reply by long1eu** [reply]
https://github.com/fluttercommunity/firebase_dart_sdk/tree/develop/cloud_firestore/cloud_firestore_dart
yes

**Comment by long1eu** [comment]
Developement for `firebase_dart_sdk` has recently resumed. What we have right now is:

* `google_sign_in`
* `firebase_auth`
* `firestore`

As of now the are a bit dated, especially firestore, but I am working to bring them back.

**Reply by Jonas-Sander** [reply]
What were the biggest advantages for you? So why are you building the native SDKs instead of using the FlutterFire ones?

**Reply by long1eu** [reply]
Except for not waiting 10 mins for ios to build firestore, the fact that they are compatible with desktop or any other platform that dart supports.

**Reply by Salakar** [reply]
@long1eu just in case you missed it, I posted this a while back: https://github.com/FirebaseExtended/flutterfire/issues/2751
 
Significantly improves Firestore build times on iOS :)

**Reply by orestesgaolin** [reply]
I think the best thing is that you might be able to run Firebase-based apps on sth like RaspberryPi or TV Box that otherwise wouldn't support it. 

Also, ability to actually debug the Firebase code is something we would love to have. The delight of being able to debug the Flutter or some packages could be extended to Firebase.

**Comment by grundid** [comment]
Having a native dart implementation in the future would be great. For now it would be enough if we could have dart APIs that all plugins/packages use. For us a firestore API would be very helpful. We have firestore dart code on flutter, angular_dart and cloud functions.

At the moment we created an API (https://github.com/dartclub/firestore_dart/tree/master/firestore_api) and wrapper for flutter and web. Our business logic depends on our on API and we initialize the wrappers on the different platforms.

**Comment by esDotDev** [comment]
Indeed it would be great to move anything that can be dart, be dart. 

The advantages are pretty obvious if you're ever tried to build the native SDK on MacOS or iOS on many different projects and machines. Out of the box it's extremely slow, and you need jump through hoops to speed it up, and even then it complicates the entire build chain, and can often lead to cryptic and hard to solve build errors. 

And then there is the obvious cross-platform conversation, where native SDK are inherently non-portable, not working everywhere Flutter does, like Amazon, PI, Windows, Linux. So this approach is very much at odds with the goals of Flutter itself.

So to recap: Native SDK's are: slow, complicated, non portable and flaky to build.

On the other hand, things written in dart, are light weight, fast to compile and 1000% portable.

To me the POV is backwards here. You're thinking native first, with dart where it makes sense. It should be flipped. Dart code strongly preferred and used wherever possible, native code only when strictly necessary, and fail gracefully when functionality doesn't exist on that platform.

As an alternative for now, we have been looking at: https://pub.dev/packages/firebase_database_rest and https://pub.dev/packages/firebase_auth_rest but of course this requires moving all our data which firestore does not make easy (who needs database backups/import anyways! MySQL only had them for 2 decades :p )

**Reply by Ehesp** [reply]
I totally agree with you, I'd love a Dart only package but it's simply not feasible (unless it's heavily invested in by Google themselves) in my opinion - even then packages like Messaging simply won't work. There are many complexities which the Firebase SDKs achieve, which Dart would have to fully replicate (think offline database caching with syncing to servers on network connection changes). It's possible for sure but it's a gigantic project. If it was also offered as a first-class SDK, it would have to be absolutely battle tested - right now a lot of the complexity is shifted to the already tried and tested existing SDKs.

> To me the POV is backwards here. You're thinking native first, with dart where it makes sense. It should be flipped. Dart code strongly preferred and used wherever possible, native code only when strictly necessary, and fail gracefully when functionality doesn't exist on that platform.

This doesn't work well though in a real-life scenario. You've got native holding onto it's own state (e.g. an authenticated user), however you then have a scenario whereby some other functionality could be integrated with Dart-only. The two still need to interact with each other, and if not, you're going to have to write a full SDK for Dart anyway. I'm not saying it's impossible, I personally just think it's not worthwhile with the extra work this approach generates on it's own.

On-top of this, you can't have a Dart-only SDK for say Cloud Functions and a native approach with Auth, since they directly work with each other (automatically authenticating requests etc).

> As an alternative for now, we have been looking at: pub.dev/packages/firebase_database_rest and pub.dev/packages/firebase_auth_rest but of course this requires moving all our data which firestore does not make easy (who needs database backups/import anyways! MySQL only had them for 2 decades :p )

I wouldn't say this is an alternative, it's a total different approach to using Firebase in your app. By using the REST APIs you lose the benefits of the native SDKs (offline, caching, performance considerations etc).

> So to recap: Native SDK's are: slow, complicated, ~~non portable and flaky to build~~.

What makes you say they're slow? I'd like to see some numbers for comparison. I disagree with you on being complicated - if you're comfortable with native code then they're fairly straightforwards.

**Comment by esDotDev** [comment]
Right, to clarify, when your needs are Firestore and Auth, which in my
experience is the core value proposition of firebase, and the 2 features I
use most often by far, then this is a nice alternative.

The frustration with this conversation from dev perspective, is we're not
necessarily interested in all those advanced features, yet they are used as
rationale why we cant get the simpler more common API we do want. This is
why a dart first approach, with native extensions that are additive seems
like a far better long term approach.

Rather than:
* You cant use Features A or B, because C is not supported.

It would be:
* You can use A or B, bit C is not supported on this platform.

From a ignorant developer perspective, the forner doesnt make sense.

In terms of slowness, I mean primarily at build time. Including the native
sdks drastically inflates your build times on ios and macos, leading
directly to a loss in revenues and productivity (not to mention a large
decrease in developer happiness/joy). There are workarounds for this, but
even when using the precompiled binaries, builds are still fairly slow, and
add yet more links in an already complicated chain, which can lead to
weeklong cleaning/building marathons with very cryptic errors. Complexity
kills, and this adds a lot.

When you are only after a basic database connection and auth this is pretty
frustrating, cause you know that under the hood, were just passing json
blobs over the wire and it could basically all be done with REST, and youre
going through all this because of features you have no desire to use
anyways.

**Comment by guyluz11** [comment]
I would love to add to the "1000% portable" part.

We talked about Linux and windows (with flutter), in my case I am using native dart with cloud firestore.
Being able to talk to firestore from native dart (without flutter) opens up new world where you can create servers and make them interact in parallel with firestore.
Now more than ever when (I think in the last key note) dart team talked about giving more place to dart servers.

In my use case the server is smart home hub/device that allow local connection control from the app, and remote connection through firestore.
This way I can get remote connection that is secure and react very feast since firestore support real-time stream. 
   
As @esDotDev said, I would love to see simplified implementation of the library and be aware that it will not get advance features that are over the head to implement like saving locally the data (not cache memory).
